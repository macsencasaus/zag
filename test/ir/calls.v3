// RUN: %vector -emit-ir %s

fn id(x: i32): i32 {
    return x;
}

fn call_id(x: i32): i32 {
    return id(x);
}

fn factorial(n: u32): u32 {
    if (n == 0)
        return 1;
    return n * factorial(n - 1);
}

fn factorial_tail(n: u32, res: u32): u32 {
    if (n == 0)
        return res;
    return factorial_tail(n - 1, n * res);
}

fn func_ptrs(func: fn(i32): i32): i32 {
    return func(2);
}
