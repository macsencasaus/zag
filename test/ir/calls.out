; V3 Readable Intermediate Representation
; Function header formatted as:
; @func[<max-stack-size>](<param-type>,...)


@id[4](i32) i32:
    ret i32 %s[0]

@call_id[8](i32) i32:
    %s[4] = i32 call @id(i32 %s[0])
    ret i32 %s[4]

@factorial[16](i32) i32:
    %s[4] = eq i32 %s[0], i32 0
    jmpz .L0, i32 %s[4]
    ret i32 1
.L0:
    %s[8] = sub i32 %s[0], i32 1
    %s[4] = i32 call @factorial(i32 %s[8])
    %s[12] = imul i32 %s[0], i32 %s[4]
    ret i32 %s[12]

@factorial_tail[20](i32, i32) i32:
    %s[8] = eq i32 %s[0], i32 0
    jmpz .L1, i32 %s[8]
    ret i32 %s[4]
.L1:
    %s[12] = sub i32 %s[0], i32 1
    %s[16] = imul i32 %s[0], i32 %s[4]
    %s[8] = i32 call @factorial_tail(i32 %s[12], i32 %s[16])
    ret i32 %s[8]

